File: src/compiler/parser/visitors/IgnorableTypedVisitor.java
package compiler.parser.visitors;

import compiler.lexer.token.CommentToken;
import compiler.lexer.token.EOFToken;
import compiler.lexer.token.WhitespaceToken;

import static compiler.lexer.token.TypeToken.VoidToken;

public interface IgnorableTypedVisitor<T> {
  T visit(CommentToken comment);

  T visit(EOFToken comment);

  T visit(WhitespaceToken comment);

  T visit(VoidToken ignored);
}

File: src/compiler/parser/visitors/IgnorableTypedVisitor.kt
package compiler.parser.visitors

import compiler.parser.Language.Token.IgnorableTokens
import compiler.parser.Language.Token.IgnorableTokens.*
import compiler.parser.Language.Token.KeywordToken.TypeToken.VoidToken


interface IgnorableTypedVisitor<T> {
    fun accept(token: IgnorableTokens) = when (token) {
        is CommentToken -> visit(token)
        is EOFToken -> visit(token)
        is WhitespaceToken -> visit(token)
    }

    fun visit(token: CommentToken): T
    fun visit(token: EOFToken): T
    fun visit(token: WhitespaceToken): T
    fun visit(token: VoidToken): T
}
